python/cucim/src/cucim/skimage/measure/_moments.py

```diff
@@ -519,20 +520,19 @@ def centroid(image, *, spacing=None):
 
 
 def _get_inertia_tensor_2x2_kernel():
+    kernel_directory = os.path.join(
+        os.path.normpath(os.path.dirname(__file__)), 'cuda'
+    )
+    with open(os.path.join(kernel_directory, "moments.h"), 'rt') as f:
+        preamble = f.read()
+
     operation = """
-    F mu0, mxx, mxy, myy;
-    mu0 = mu[0];
-    mxx = mu[6];
-    myy = mu[2];
-    mxy = mu[4];
-
-    result[0] = myy / mu0;
-    result[1] = result[2] = -mxy / mu0;
-    result[3] = mxx / mu0;
+    inertia_tensor_2x2(&mu[0], &result[0]);
```

Comment:
FWICT CuPy actually passes [`CArray`]( https://github.com/cupy/cupy/blob/aa12407137ae15ad014e14842c5e2a18338716cd/cupy/_core/include/cupy/carray.cuh#L379 ) objects into the kernel. These can be [indexed]( https://github.com/cupy/cupy/blob/aa12407137ae15ad014e14842c5e2a18338716cd/cupy/_core/include/cupy/carray.cuh#L588 ) though, which in [the C-contiguous case]( https://github.com/cupy/cupy/blob/aa12407137ae15ad014e14842c5e2a18338716cd/cupy/_core/include/cupy/carray.cuh#LL589 ) indexes [directly into the `data_` pointer]( https://github.com/cupy/cupy/blob/aa12407137ae15ad014e14842c5e2a18338716cd/cupy/_core/include/cupy/carray.cuh#L592 ). We leverage this here to back out the pointer to underlying array data, which the kernel then takes for normal operation.

Reply:
as long as we make sure to check that the inputs are C-contiguous, this seems okay

Reply:
There is a fallback for [the non-contiguous case]( https://github.com/cupy/cupy/blob/aa12407137ae15ad014e14842c5e2a18338716cd/cupy/_core/include/cupy/carray.cuh#L593 )

That said, the C code effectively assumes the data is contiguous somehow. We could require that before calling the kernel. Alternatively we could rewrite the kernel so it is less reliant on this requirement. Though not sure how often more complex cases are needed given this seems to be an internal function. Please feel free to correct me if I'm missing something